client/relay-chain-rpc-interface/src/rpc_client.rs

```diff
@@ -0,0 +1,478 @@
+	},
+	InboundDownwardMessage, ParaId, PersistedValidationData,
+};
+use cumulus_relay_chain_interface::{RelayChainError, RelayChainResult};
+use futures::{
+	channel::mpsc::{Receiver, Sender},
+	StreamExt,
+};
+use jsonrpsee::{
+	core::{
+		client::{Client as JsonRpcClient, ClientT, Subscription, SubscriptionClientT},
+		Error as JsonRpseeError,
+	},
+	rpc_params,
+	types::ParamsSer,
+	ws_client::WsClientBuilder,
+};
+use parity_scale_codec::{Decode, Encode};
+use polkadot_service::TaskManager;
+use sc_client_api::StorageData;
+use sc_rpc_api::{state::ReadProof, system::Health};
+use sp_core::sp_std::collections::btree_map::BTreeMap;
+use sp_runtime::DeserializeOwned;
+use sp_storage::StorageKey;
+use std::sync::Arc;
+use tokio::sync::mpsc::{
+	channel as tokio_channel, Receiver as TokioReceiver, Sender as TokioSender,
+};
+
+pub use url::Url;
+
+const LOG_TARGET: &str = "relay-chain-rpc-client";
+
+const NOTIFICATION_CHANNEL_SIZE_LIMIT: usize = 20;
+
+/// Client that maps RPC methods and deserializes results
+#[derive(Clone)]
+pub struct RelayChainRpcClient {
+	/// Websocket client to make calls
+	ws_client: Arc<JsonRpcClient>,
+
+	/// Retry strategy that should be used for requests and subscriptions
+	retry_strategy: ExponentialBackoff,
+
+	to_worker_channel: Option<TokioSender<NotificationRegisterMessage>>,
+}
+
+/// Worker messages to register new notification listeners
+#[derive(Clone, Debug)]
+pub enum NotificationRegisterMessage {
+	RegisterBestHeadListener(Sender<PHeader>),
+	RegisterImportListener(Sender<PHeader>),
+	RegisterFinalizationListener(Sender<PHeader>),
+}
+
+/// Worker that should be used in combination with [`RelayChainRpcClient`]. Must be polled to distribute header notifications to listeners.
+struct RpcStreamWorker {
+	// Communication channel with the RPC client
+	client_receiver: TokioReceiver<NotificationRegisterMessage>,
+
+	// Senders to distribute incoming header notifications to
+	imported_header_listeners: Vec<Sender<PHeader>>,
+	finalized_header_listeners: Vec<Sender<PHeader>>,
+	best_header_listeners: Vec<Sender<PHeader>>,
+
+	// Incoming notification subscriptions
+	rpc_imported_header_subscription: Subscription<PHeader>,
+	rpc_finalized_header_subscription: Subscription<PHeader>,
+	rpc_best_header_subscription: Subscription<PHeader>,
+}
+
+/// Entry point to create [`RelayChainRpcClient`] and start a worker that distributes notifications.
+pub async fn create_client_and_start_worker(
+	url: Url,
+	task_manager: &mut TaskManager,
+) -> RelayChainResult<RelayChainRpcClient> {
+	let mut client = RelayChainRpcClient::new(url).await?;
+	let best_head_stream = client.subscribe_new_best_heads().await?;
+	let finalized_head_stream = client.subscribe_finalized_heads().await?;
+	let imported_head_stream = client.subscribe_imported_heads().await?;
+
+	let (worker, sender) =
+		RpcStreamWorker::new(imported_head_stream, best_head_stream, finalized_head_stream);
+	client.set_worker_channel(sender);
```

Comment:
Why don't we first create the worker and then the client? Then we don't need to make the `sender` an `Option`.

Reply:
Worker took in the streams which required the client. I changed the methods to be static, so now we got rid of the `Option` 👍